### PR TITLE
Allow specifying "onPolicyCreationFailure" mode for trusted types

### PR DESCRIPTION
**Pull request by Zlatkovsky**
Webpack already allows for specifying a trusted-types policy name. However, its current implementation is such that if a call to `trustedTypes.createPolicy` fails, the code will _immediately stop executing_. This isn't necessarily desirable, as an application could be in the early phases of rolling out trusted types, and thus have the CSP rule for `trusted-types LibraryA LibraryB etc`, BUT have `require-trusted-types-for 'script'` be in "report only" mode (`Content-Security-Policy-Report-Only`). In such a configuration, and when the webpacked code is dynamically-loaded into an application, adding the policy name to the webpack config will break old versions.

This PR keeps the original behavior, but introduces a new option for `onPolicyCreationFailure: "continue" | "stop"` (with "stop" remaining the default). If a developer chooses the "continue" option, the policy-creation will be wrapped in a try/catch. There is no security risk to this, since for host applications that DO have strict enforcement of trusted-types, the code will simply fail when the dangerous sink is used (e.g., when doing `parseFromString`). And likewise, wrapping in try/catch and doing nothing on catch is OK, because the code already deals with the possibility of the trustedTypes API not being available on the browser.

**An example of the behaviors -- with and without the new option, in both "strict" and "report-only" modes -- can be found in a demo app here: <https://webpack-trusted-types-example.azurewebsites.net/>**

For testing: I added two more tests that show:
1) The code indeed stopping on policy-creation failure in the default case.
2) The code proceeding (and just showing a warning) if the `onPolicyCreationFailure: "continue"` is specified.

It's worth noting that the practice of wrapping policy-creation in a try/catch is generally accepted. For example, [DOMPurify](https://www.npmjs.com/package/dompurify), a wildly popular trusted-types packages, has the following code: <https://github.com/cure53/DOMPurify/blob/1c63c4b5a7da8c036dfc32e148f1042d4a77c18e/src/purify.js#L68>

DOMPurify try-catch

### <samp>🤖 Generated by Copilot at 2c51eb6</samp>

This pull request adds a new `onPolicyCreationFailure` option to the `trustedTypes` output configuration, which allows users to control how the webpack runtime handles the case when creating a trusted types policy fails. It also adds tests and documentation for the new option, and removes the BOM from a test file.

### <samp>🤖 Generated by Copilot at 2c51eb6</samp>

*  Add a new option `onPolicyCreationFailure` to the `trustedTypes` output configuration, which controls how the runtime handles the case when creating a trusted types policy fails ([link](https://github.com/webpack/webpack/pull/16990/files?diff=unified&w=0#diff-c0078e4713a0e7a24dbc56d6dfda3cf139376325fc65ea70a8389d4d30b27f04R964), [link](https://github.com/webpack/webpack/pull/16990/files?diff=unified&w=0#diff-0cdb656a2796a80418de2964257360de642bd5a770102d52366890a65aa912a0R5008-R5011))
*  Modify the runtime code for creating and using trusted types policies to use the `onPolicyCreationFailure` option, and wrap the policy creation in a `try-catch` block if the option is `"continue"` ([link](https://github.com/webpack/webpack/pull/16990/files?diff=unified&w=0#diff-eab2f2ec28ba995308ad37351b95f00b753b37cc97769432b9ade097b8ad77abR28-R30), [link](https://github.com/webpack/webpack/pull/16990/files?diff=unified&w=0#diff-eab2f2ec28ba995308ad37351b95f00b753b37cc97769432b9ade097b8ad77abL61-R82))
*  Remove the byte order mark (BOM) from the `test/cases/parsing/bom/bomfile.js` file, which is used to test the parsing of files with BOM ([link](https://github.com/webpack/webpack/pull/16990/files?diff=unified&w=0#diff-c4230b2fca9242c7e0a4582b6fc2600630032c52e18ea9a2eab8dbdc7d76a767L1-L0))
*  Add new test cases to `test/configCases/trusted-types` to test the behavior of the `onPolicyCreationFailure` option when set to `"continue"` or `"stop"` ([link](https://github.com/webpack/webpack/pull/16990/files?diff=unified&w=0#diff-6401535d9dbfa6584111e134c12b3d28d376c01bc4db845acd9d3941c2a3809dR1-R36), [link](https://github.com/webpack/webpack/pull/16990/files?diff=unified&w=0#diff-84989f397236ba3fba047d76f5da69e95c7c019fa58dd19b544f78400c9ec0ebR1-R17), [link](https://github.com/webpack/webpack/pull/16990/files?diff=unified&w=0#diff-5deb01296d1f1aa1c6675ae4026d03b93ca75a46a9af419e963c2b464e967586R1-R38), [link](https://github.com/webpack/webpack/pull/16990/files?diff=unified&w=0#diff-5e3bdc992574b84f6ca6178195dd9d73708d180c224084ca2485a976213c578eR1-R14))

### CODE REVIEW

**Review by alexander-akait — COMMENTED** [review]
Can you add real test case in our config tests?

**Comment by Zlatkovsky** [comment]
> Can you add real test case in our config tests?

@alexander-akait , can you point me at an example of what you mean? I did add tests to `configCases/trusted-types`, is there another place I should add to as well?

image

File: test/cases/parsing/bom/bomfile.js
Comment (alexander-akait) [inline test/cases/parsing/bom/bomfile.js:1]: There is BOM, don't touch file file, it is a test case

File: test/configCases/trusted-types/continue-on-policy-creation-failure/index.js
Comment (alexander-akait) [inline test/configCases/trusted-types/continue-on-policy-creation-failure/index.js:10]: I mean please create a real test case (as you describe in your PR description), `spy` is like unit testing, that is not good

**Comment by alexander-akait** [comment]
Also you need accept CLA, can you do it?

File: test/configCases/trusted-types/continue-on-policy-creation-failure/index.js
Comment (Zlatkovsky) [reply test/configCases/trusted-types/continue-on-policy-creation-failure/index.js:10]: @alexander-akait , I'm not sure that there's a way to add a check-in-able test case -- in order to see the effect, you need to have a website running and serving CSP headers. But I can work on making a simple repro app, I'll try to do that later today or tomorrow.

Comment (alexander-akait) [reply test/configCases/trusted-types/continue-on-policy-creation-failure/index.js:10]: Yeah thank you

**Comment by Zlatkovsky** [comment]
@alexander-akait , apologies for the delay. I have signed the CLA, and also finished creating a demo app that showcases the behaviors -- with and without the new option, in both "strict" and "report-only" modes -- here: <https://webpack-trusted-types-example.azurewebsites.net/>

I see the branch has a conflict in the auto-generated WebpackOptions.check.js file. I can merge with latest and generate a new `schemas/WebpackOptions.check.js` file as soon as I get the green light from you.

(Also, should I bump the package.json as part of my PR, or is that something that the repo maintainers and/or a bot does)?

Thanks so much!

**Comment by alexander-akait** [comment]
@Zlatkovsky Thank you for link, I see, feel fre to rebase

**Comment by Zlatkovsky** [comment]
@alexander-akait , thanks for your review! I've rebased and pushed a new commit.

**Comment by alexander-akait** [comment]
Can you update `Defaults.unittest.js`, snapshots?

**Comment by Zlatkovsky** [comment]
@alexander-akait , yep, just updated it. (Sorry, didn't see the failure earlier).

**Comment by alexander-akait** [comment]
> Another thing I saw was that some of the codecoverage checks were failing. Can you point me to what I can do about those?

Ignore it, we fixed it, but I think you have used the old version of the `main` branch in your fork

**Comment by Zlatkovsky** [comment]
@alexander-akait , even after merging with the latest, it looks like I still have one failing `codecov/patch/integration` check, as well as a `Test / integration (ubuntu-latest, 10.x, a)` that must have timed out; it's running for 4+ hours now.

Please let me know what, if anything, I can do to help resolve these; and in general, if there's anything further that I can do to get this PR in :-)

Thanks!

**Comment by alexander-akait** [comment]
@Zlatkovsky Yeah, just ignore, last time github actions are not stable, 99% our fails due network issue with codecov/node source code repo/etc services

**Review by TheLarkInn — APPROVED** [review]
Congrats @Zlatkovsky on your first webpack contribution! Shipping today. 🏅

**Pull request merged**
